services/w3c/w3c-validation.service.js

```diff
@@ -0,0 +1,118 @@
+'use strict'
+const Joi = require('@hapi/joi')
+const { optionalUrl } = require('../validators')
+const {
+  documentation,
+  presetRegex,
+  getColor,
+  getMessage,
+  getSchema,
+} = require('./w3c-validation-helper')
+const { BaseJsonService, NotFound } = require('..')
+
+const schema = Joi.object({
+  messages: Joi.array()
+    .required()
+    .items(
+      Joi.object({
+        type: Joi.string()
+          .allow('info', 'error', 'non-document-error')
+          .required(),
+        message: Joi.string().required(),
+      })
+    ),
+}).required()
+
+const queryParamSchema = Joi.object({
+  targetUrl: optionalUrl.required(),
+  preset: Joi.string()
+    .regex(presetRegex)
+    .allow(''),
+}).required()
+
+module.exports = class W3cValidation extends BaseJsonService {
+  static get category() {
+    return 'analysis'
+  }
+
+  static get route() {
+    return {
+      base: 'w3c-validation',
+      pattern: ':parser(default|html|xml|xmldtd)',
+      queryParamSchema,
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'W3C Validation',
+        namedParams: { parser: 'html' },
+        queryParams: {
+          targetUrl: 'https://validator.nu/',
+          preset: 'HTML, SVG 1.1, MathML 3.0',
+        },
+        staticPreview: this.render({ messageTypes: {} }),
+        documentation,
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'w3c',
+    }
+  }
+
+  static render({ messageTypes }) {
+    return {
+      message: getMessage(messageTypes),
+      color: getColor(messageTypes),
+    }
+  }
+
+  async fetch(targetUrl, preset, parser) {
+    return this._requestJson({
+      url: 'https://validator.nu/',
+      schema,
+      options: {
+        qs: {
+          schema: getSchema(preset),
+          parser: parser === 'default' ? undefined : parser,
+          doc: encodeURI(targetUrl),
+          out: 'json',
+        },
+      },
+    })
+  }
+
+  transform(messages) {
+    return messages.reduce((accumulator, message) => {
+      let { type } = message
+      if (type === 'non-document-error') {
```

Comment:
Is this check sufficient on its own? Is it possible to have a `non-document-error` that's _not_ due to an invalid/inaccessible url for the `doc` param?

Reply:
Not sure if there might be others but the last 4 was the frameset error. That was fixed with the url change. The other option will be to do some form or regex check for the 404 message and another regex for the `"adfasdfasdfasdfadfadfadfasdfadf.com: Name or service not known"` check

Reply:
I suppose it is a matter of what the response looks like from the upstream API when the target url cannot be found, and under what circumstances that API response may include an object in `messages` array that has the `non-document-error` type.

My sense from the little bit of experimentation I've done is that we'll get a response with a single object in `messages` that will look something like this:

```json
{
  "messages": [
    {
      "type": "non-document-error",
      "subType": "io",
      "message": "{{target url domain name}}: Name or service not known"
    }
  ]
}
```

What I'm wondering is if we could ever see a response from the API that looks something like this:
```json
{
  "messages": [
    {
      "type": "non-document-error",
      "subType": "{{something other than io}}",
      "message": "{{some other type of message that's not a 404 on the url for doc"
    },
    {
      "type": "error",
      "message": "Element “meta” is missing one or more of the following attributes..."
    }
  ]
}
```
Because if so then I suspect it would be pretty confusing for users to see the target url not found badge ![](https://img.shields.io/badge/w3c-target%20url%20not%20found-red.svg)

Based on what you've seen so far, do you think that's a possible scenario? Or if the `messages` array has _any_ `non-document-error` type messages can we safely assume that's a target url "not found" scenario?

Reply:
I did a search through the codebase at `https://github.com/validator/validator`, from a high level view. I think implementing the usual 404 might not be a good idea like what you mention. Reading the documentation in detail as well. There is a high chance that there are more types of non-document-error besides those indicating that the url is not valid

```
"non-document-error" signifies an error that causes the checking to end in an indeterminate state because the document being validated could not be examined to the end. Examples of such errors include broken schemas, bugs in the validator and IO errors. (Note that when a schema has parse errors, they are first reported as errors and then a catch-all non-document-error is also emitted.)
```

Reply:
Thanks for digging into it. My takeaway is that we need to do some additional checks to more accurately detect the target url not found scenario. I'd suggest the following:

* include `subType` and `message` in the schema on the objects in the messages array so that we'll have access to those values
* In the `transform` function, include the following conditions to better detect target url not found: 
    1. (optional) check `messages.length` for a single message. my assumption is that if there's an io error with the target url (specified via the `doc` param) that there wouldn't be any other messages
   2. check `type` for `non-document-error`
   3. check `subType` for `io`
   4. check `message` for a `Name or service not known` suffix
   5. (optional) check the full `message`, expecting something like `{{targetUrl with http(s) and trailing paths stripped}}: Name or service not known` For example: if the targetUrl was `https://notrealgithub.com/foo` then `message` should be `notrealgithub.com: Name or service not known`

Reply:
Sorry for the radio silence had to swing back to work on a Nanodegree. I made the notfound tests stricter.